File: scripts/generate_catalog.py
#!/usr/bin/env python3
"""Generate llms.txt and docs/skills-catalog.md from SKILL.md frontmatter.

Reads name + description from each `.agent-src/skills/*/SKILL.md` and writes:

- `llms.txt`            — machine-readable index (plain text)
- `docs/skills-catalog.md` — human-readable catalog (markdown)

Idempotent. Safe to re-run. Sort order: alphabetical by skill name.

Usage:
    python3 scripts/generate_catalog.py
"""

from __future__ import annotations

import re
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
SKILLS_DIR = REPO_ROOT / ".agent-src" / "skills"
LLMS_TXT = REPO_ROOT / "llms.txt"
CATALOG_MD = REPO_ROOT / "docs" / "skills-catalog.md"

FRONTMATTER_RE = re.compile(r"^---\s*\n(.*?)\n---", re.DOTALL)
NAME_RE = re.compile(r"^name:\s*(.+?)\s*$", re.MULTILINE)
DESC_RE = re.compile(r"^description:\s*\"?(.+?)\"?\s*$", re.MULTILINE)


def parse_skill(skill_md: Path) -> tuple[str, str] | None:
    text = skill_md.read_text(encoding="utf-8")
    m = FRONTMATTER_RE.match(text)
    if not m:
        return None
    front = m.group(1)
    name_m = NAME_RE.search(front)
    desc_m = DESC_RE.search(front)
    if not name_m or not desc_m:
        return None
    return name_m.group(1).strip(), desc_m.group(1).strip()


def collect_skills() -> list[tuple[str, str]]:
    skills = []
    for skill_dir in sorted(SKILLS_DIR.iterdir()):
        if not skill_dir.is_dir():
            continue
        skill_md = skill_dir / "SKILL.md"
        if not skill_md.exists():
            continue
        parsed = parse_skill(skill_md)
        if parsed is not None:
            skills.append(parsed)
    return skills


def render_llms_txt(skills: list[tuple[str, str]]) -> str:
    lines = [
        "# agent-config — Skill Index",
        "",
        "Machine-readable index of all skills in this package. Each line:",
        "  <skill-name>: <one-line description>",
        "",
        "Source: .agent-src/skills/<name>/SKILL.md",
        "Catalog: docs/skills-catalog.md",
        "",
    ]
    for name, desc in skills:
        lines.append(f"{name}: {desc}")
    lines.append("")
    return "\n".join(lines)


def render_catalog_md(skills: list[tuple[str, str]]) -> str:
    lines = [
        "# Skills Catalog",
        "",
        f"All **{len(skills)} skills** available in this package, in alphabetical order.",
        "Click a skill name to open its SKILL.md and read the full guidance.",
        "",
        "> **Regenerate:** `python3 scripts/generate_catalog.py`",
        "> This file is auto-generated from `SKILL.md` frontmatter — do not edit manually.",
        "",
        "| Skill | What your agent learns |",
        "|---|---|",
    ]
    for name, desc in skills:
        lines.append(f"| [`{name}`](../.agent-src/skills/{name}/SKILL.md) | {desc} |")
    lines.append("")
    lines.append("---")
    lines.append("")
    lines.append("← [Back to README](../README.md)")
    lines.append("")
    return "\n".join(lines)


def main() -> int:
    if not SKILLS_DIR.exists():
        print(f"❌  Skills directory not found: {SKILLS_DIR}")
        return 1
    skills = collect_skills()
    if not skills:
        print("❌  No skills found.")
        return 1

    LLMS_TXT.write_text(render_llms_txt(skills), encoding="utf-8")
    CATALOG_MD.parent.mkdir(parents=True, exist_ok=True)
    CATALOG_MD.write_text(render_catalog_md(skills), encoding="utf-8")

    print(f"✅  Wrote {LLMS_TXT.relative_to(REPO_ROOT)} ({len(skills)} skills)")
    print(f"✅  Wrote {CATALOG_MD.relative_to(REPO_ROOT)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/generate_catalog.ts
#!/usr/bin/env node
/**
 * Generate llms.txt and docs/skills-catalog.md from SKILL.md frontmatter.
 *
 * TypeScript twin of `src/scripts/generate_catalog.py` (ADR-089, Phase 8
 * Wave 8a). Mirrors the Python CLI contract EXACTLY — same stdout lines,
 * same exit codes, byte-identical generated `llms.txt` + `docs/skills-catalog.md`.
 *
 * Reads name + description from each `dist/agent-src/skills/*\/SKILL.md` and
 * writes:
 *
 * - `llms.txt`            — machine-readable index (plain text)
 * - `docs/skills-catalog.md` — human-readable catalog (markdown)
 *
 * Idempotent. Safe to re-run. Sort order: alphabetical by skill name.
 *
 * Usage:
 *     node scripts/generate_catalog.ts
 */
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const _HERE = path.dirname(fileURLToPath(import.meta.url));
// _HERE === <repo>/src/scripts ; parent.parent.parent of the .py file
// (src/scripts/generate_catalog.py) is the repo root — two dirs up from scripts.
export const REPO_ROOT = path.resolve(_HERE, '..', '..');
export const SKILLS_DIR = path.join(REPO_ROOT, 'dist/agent-src', 'skills');
export const LLMS_TXT = path.join(REPO_ROOT, 'llms.txt');
export const CATALOG_MD = path.join(REPO_ROOT, 'docs', 'skills-catalog.md');

// re.compile(r"^---\s*\n(.*?)\n---", re.DOTALL) anchored with .match (start only).
// JS \s matches newlines like Python's, so `\s*\n` and `(.*?)` (DOTALL → [\s\S]).
const FRONTMATTER_RE = /^---\s*\n([\s\S]*?)\n---/;
// re.compile(r"^name:\s*(.+?)\s*$", re.MULTILINE)
const NAME_RE = /^name:[^\S\n]*(.+?)[^\S\n]*$/m;
// re.compile(r"^description:\s*\"?(.+?)\"?\s*$", re.MULTILINE)
const DESC_RE = /^description:[^\S\n]*"?(.+?)"?[^\S\n]*$/m;

/** Mirror Python str.strip() — strip ASCII + Unicode whitespace from both ends. */
function pyStrip(s: string): string {
    return s.replace(/^\s+/, '').replace(/\s+$/, '');
}

export function parse_skill(skill_md: string): [string, string] | null {
    const text = fs.readFileSync(skill_md, 'utf-8');
    const m = FRONTMATTER_RE.exec(text);
    // `.match` semantics: must match at start of string.
    if (!m || m.index !== 0) {
        return null;
    }
    const front = m[1] as string;
    const name_m = NAME_RE.exec(front);
    const desc_m = DESC_RE.exec(front);
    if (!name_m || !desc_m) {
        return null;
    }
    return [pyStrip(name_m[1] as string), pyStrip(desc_m[1] as string)];
}

/** Sorted immediate child entries (mirrors `sorted(SKILLS_DIR.iterdir())`). */
function _iterdirSorted(p: string): string[] {
    let names: string[];
    try {
        names = fs.readdirSync(p);
    } catch {
        return [];
    }
    names.sort();
    return names.map((name) => path.join(p, name));
}

function _isDir(p: string): boolean {
    try {
        return fs.statSync(p).isDirectory();
    } catch {
        return false;
    }
}

function _exists(p: string): boolean {
    try {
        fs.statSync(p);
        return true;
    } catch {
        return false;
    }
}

export function collect_skills(): Array<[string, string]> {
    const skills: Array<[string, string]> = [];
    for (const skill_dir of _iterdirSorted(SKILLS_DIR)) {
        if (!_isDir(skill_dir)) {
            continue;
        }
        const skill_md = path.join(skill_dir, 'SKILL.md');
        if (!_exists(skill_md)) {
            continue;
        }
        const parsed = parse_skill(skill_md);
        if (parsed !== null) {
            skills.push(parsed);
        }
    }
    return skills;
}

export function render_llms_txt(skills: Array<[string, string]>): string {
    const lines = [
        '# agent-config — Skill Index',
        '',
        'Machine-readable index of all skills in this package. Each line:',
        '  <skill-name>: <one-line description>',
        '',
        'Source: dist/agent-src/skills/<name>/SKILL.md',
        'Catalog: docs/skills-catalog.md',
        '',
    ];
    for (const [name, desc] of skills) {
        lines.push(`${name}: ${desc}`);
    }
    lines.push('');
    return lines.join('\n');
}

export function render_catalog_md(skills: Array<[string, string]>): string {
    const lines = [
        '# Skills Catalog',
        '',
        `All **${skills.length} skills** available in this package, in alphabetical order.`,
        'Click a skill name to open its SKILL.md and read the full guidance.',
        '',
        '> **Regenerate:** `python3 scripts/generate_catalog.py`',
        '> This file is auto-generated from `SKILL.md` frontmatter — do not edit manually.',
        '',
        '| Skill | What your agent learns |',
        '|---|---|',
    ];
    for (const [name, desc] of skills) {
        lines.push(`| [\`${name}\`](../dist/agent-src/skills/${name}/SKILL.md) | ${desc} |`);
    }
    lines.push('');
    lines.push('---');
    lines.push('');
    lines.push('← [Back to README](../README.md)');
    lines.push('');
    return lines.join('\n');
}

/** POSIX relative path of `child` under `root` (mirrors `relative_to().as_posix()`). */
function _relToRepo(child: string): string {
    return path.relative(REPO_ROOT, child).split(path.sep).join('/');
}

export function main(): number {
    if (!_exists(SKILLS_DIR)) {
        process.stdout.write(`❌  Skills directory not found: ${SKILLS_DIR}\n`);
        return 1;
    }
    const skills = collect_skills();
    if (skills.length === 0) {
        process.stdout.write('❌  No skills found.\n');
        return 1;
    }

    fs.writeFileSync(LLMS_TXT, render_llms_txt(skills), 'utf-8');
    fs.mkdirSync(path.dirname(CATALOG_MD), { recursive: true });
    fs.writeFileSync(CATALOG_MD, render_catalog_md(skills), 'utf-8');

    process.stdout.write(`✅  Wrote ${_relToRepo(LLMS_TXT)} (${skills.length} skills)\n`);
    process.stdout.write(`✅  Wrote ${_relToRepo(CATALOG_MD)}\n`);
    return 0;
}

const _isMain = import.meta.url === pathToFileURL(path.resolve(process.argv[1] ?? '')).href;
if (_isMain) {
    process.exitCode = main();
}
